affirm/src/main/java/com/affirm/android/VcnCheckoutWebViewClient.java

```diff
@@ -20,6 +24,10 @@
     private final String receiveReasonCodes;
     private static final String VCN_CHECKOUT_REGEX = "data=";
     private static final String ENCODING_FORMAT = "UTF-8";
+    private static final String INVALID_CHECKOUT_REDIRECT_URL =
+            "https://sandbox.affirm.com/u/";
```

Comment:
Why do you have the server url hard-coded here? Specially sandbox.

Reply:
@kevindavmoore Thanks for the reminder, it has been fixed, but I am not sure about the invalid checkout redirect url of the prod and the CA env, can you help confirm it?

https://github.com/Affirm/affirm-merchant-sdk-android/blob/2e89ef59864b48f59199dabd42cdbe4a7be88e73/affirm/src/main/java/com/affirm/android/AffirmConstants.java#L116

Reply:
I am not that familiar with those, sorry. I'm wondering if it matters if it is just a redirect that you catch yourself.

Reply:
This is just a workaround, I capture this url and then exit the SDK so that the merchant can re-initiate checkout
This is the logic in JS, needs the help of front-end colleagues

Reply:
> Addressing: https://trello.com/c/y4ebrfFh/22-optional-shipping-address-bug
> 
> I created a checkout model that is exactly the same as `snippet.java`, and then used the same public_api_key as merchant, and this problem was not reproduced.
> 
> From the obtained logs, I found that the checkout status is invalid.
> 
> @kevindavmoore @AmyBeall as a part of this investigation, can you help me check the status of the checkout on the server, and the specific reason checkout status returned invalid
> 
> We got the redirect_url through the api (/api/v2/checkout/)
> https://sandbox.affirm.com/apps/mup/checkout/2PLD0V5QP0GA2KIH/new/47IJQMSZM8E02WCH/
> 
> Then WebView loads this url, and JS get checkout data
> https://sandbox.affirm.com/api/v1/app_configurations/checkout/47IJQMSZM8E02WCH?public_api_key=2PLD0V5QP0GA2KIH
> {
> "Message": "Checkout status is in an invalid state.",
> "Redirect_url": "https://sandbox.affirm.com/u/"
> }
> 
> In this case, I think the SDK can dismiss the WebView first, and then return an error message to the merchant. That user can re-initiate checkout.

Hello!

I checked the logs and found something odd. Each time a checkout is run it should generate a new checkout token. The requests above is using the following checkout token : 47IJQMSZM8E02WCH. This checkout token has been used numerous times by the SDK which is why the error is being returned. Looking at all of the requests from the SDK for this merchant, this particular checkout token is being used 50% of the time. We have never seen this before, is it possible that SDK is somehow caching the checkout token?

Reply:
@AmyBeall Thank you for your reply!
I thoroughly checked the code, but i didn't do any cache(include okhttp & checkout response/ checkout token) in SDK.
https://github.com/Affirm/affirm-merchant-sdk-android/blob/1da83b17132b498d51e33ed777a52732ded19f92/affirm/src/main/java/com/affirm/android/AffirmPlugins.java#L117

https://github.com/Affirm/affirm-merchant-sdk-android/blob/1da83b17132b498d51e33ed777a52732ded19f92/affirm/src/main/java/com/affirm/android/CheckoutRequest.java#L121

Is it normal for the merchant to use our sample? Has the merchant made any changes to the SDK? And is there any more other informations from merchant?

Reply:
I will follow up with them and get back to you. Thank you!